packages/@aws-cdk/cicd/README.md

```diff
@@ -0,0 +1,112 @@
+  ┗━━━━━━━━━━━━━━━━┛  ┗━━━━━━━━━━━━━━━━┛  ┗━━━━━━━━━━━━━━━━┛  ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+```
+
+#### `index.ts`
+```ts
+import codebuild = require('@aws-cdk/aws-codebuild');
+import codepipeline = require('@aws-cdk/aws-codepipeline');
+import cdk = require('@aws-cdk/cdk');
+import cicd = require('@aws-cdk/cicd');
+
+const app = new cdk.App();
+
+// We define a stack that contains the CodePipeline
+const pipelineStack = new cdk.Stack(app, 'CodePipelineStack');
+const pipeline = new codepipeline.Pipeline(pipelineStack, 'CodePipeline', {
+  restartExecutionOnUpdate: true,
+  /* ... */
+});
+// Configure the CodePipeline source - where your CDK App's source code is hosted
+const source = new codepipeline.GitHubSourceAction(pipelineStack, 'GitHub', {
+  stage: pipeline.addStage('source'),
+  /* ... */
+});
+const project = new codebuild.Project(pipelineStack, 'CodeBuild', {
+  artifacts: new codebuild.CodePipelineArtifacts(),
+  source: new codebuild.CodePipelineSource(),
+  /* ... */
+});
+const synthesizedApp = project.outputArtifact;
+
+// Optionally, self-update the pipeline stack
+const selfUpdateStage = pipeline.addStage('SelfUpdate');
+new cicd.PipelineDeployStackAction(pipelineStack, 'SelfUpdatePipeline', {
+  stage: selfUpdateStage,
+  stack: pipelineStack,
+  inputArtifact: synthesizedApp,
+});
+
+// Now add our service stacks
+const deployStage = pipeline.addStage('Deploy');
+const serviceStackA = new MyServiceStackA(app, 'ServiceStackA', { /* ... */ });
+const serviceStackB = new MyServiceStackB(app, 'ServiceStackB', { /* ... */ });
+// Add actions to deploy the stacks in the deploy stage:
+new cicd.PipelineDeployStackAction(pipelineStack, 'DeployServiceStack', {
+  stage: deployStage,
+  stack: serviceStackA,
+  inputArtifact: synthesizedApp,
+  createChangeSetRunOrder: 1,
+  executeChangeSetRunOrder: 2,
+});
+new cicd.PipelineDeployStackAction(pipelineStack, 'DeployServiceStack', {
+  stage: deployStage,
+  stack: serviceStackB,
+  inputArtifact: synthesizedApp,
+  createChangeSetRunOrder: 998,
+  executeChangeSetRunOrder: 999,
+});
+```
+
+#### `buildspec.yml`
+In order for the *CodeBuild* step to produce an input that meets the expectations of the `PipelineDeployStackAction`
+contract, the following `buildspec.yml` file needs to be placed that the root of the repository defined at the `Source`
+stage:
+
+```yml
+version: 0.2
+phases:
+  install:
+    commands:
+      # Installs the npm dependencies as defined by the `package.json` file
+      # present in the root directory of the package
+      # (`cdk init app --language=typescript` would have created one for you)
+      - npm install
+  build:
+    commands:
+      # Builds the CDK App so it can be synthesized
+      - npm run build
+      # Synthesizes the CDK App and puts the resulting artifacts into `dist`
+      - npm run cdk synth -- -o dist
+artifacts:
+  # The output artifact is all the files in the `dist` directory
+  base-directory: dist
+  files: '**/*'
+```
```

Comment:
I don't like this experience. `Project` in `codebuild` can also use an inline build spec. Should we put this logic in some reusable utility methods that encapsulate this logic for the languages the CDK supports? A separate method for TypeScript / JavaScript, Java, .NET, etc.

These can be 'fallback' instructions.

Reply:
Yeah I guess I can provide some canned buildspecs. I'll have a look into this.

Reply:
Looking into the canned buildspecs, I realized that it gets awkward once the build image needs to support the buildspec, especially when the standard CodeBuild images don't ... For example, if I want to make a buildspec for a Java CDK project, I'd need to be able to provide an image that has `node` and a JDK... Right now I'd have to limit support to TypeScript & Javascript, which feels too restrictive.